src/main/java/net/snowflake/client/jdbc/SnowflakeDatabaseMetaDataResultSet.java

```diff
@@ -111,23 +111,36 @@ private SnowflakeDatabaseMetaDataResultSet(
          metadataType.getColumnTypes(), rows, statement);
   }
 
+  /**
+   * Raises SQLException if the result set is closed
+   *
+   * @throws SQLException if the result set is closed.
+   */
+  private void raiseSQLExceptionIfResultSetIsClosed() throws SQLException
+  {
+    if (isClosed())
+    {
+      throw new SnowflakeSQLException(ErrorCode.RESULTSET_ALREADY_CLOSED);
+    }
+  }
+
+  @Override
+  public boolean isClosed() throws SQLException
+  {
+    return statement == null;
+  }
+
   @Override
   public boolean next() throws SQLException
   {
     logger.debug("public boolean next()");
-
+    // no exception is raised even after the result set is closed.
     if (row < rows.length)
     {
       nextRow = rows[row++];
       return true;
     }
 
-    if (statement != null)
-    {
-      statement.close();
-      statement = null;
-    }
-
```

Comment:
not closing the statement after the last next call.
This is required to avoid SQLException from next.